### PR TITLE
[dist_optim] add distributed functional Adam optimizer

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50630 [dist_optim] add warning to distributed optimizer
* #50620 [dist_optim] add distributed functional AdamW optimizer
* #50619 [dist_optim] add distributed functional RMSprop optimizer
* #50623 [dist_optim] add distributed functional Adadelta optimizer
* #50618 [dist_optim] add distributed functional sgd optimizer
* **#50624 [dist_optim] add distributed functional Adam optimizer**

Add TorchScript compatible Adam functional optimizer to distributed optimizer

Differential Revision: [D25932770](https://our.internmc.facebook.com/intern/diff/D25932770)

### CODE REVIEW

File: torch/distributed/optim/functional_adam.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+from typing import List, Dict, Optional, Tuple
+import torch
+import torch.optim.functional as F
+
+from torch import Tensor
+
+# Define a TorchScript compatible Functional Adam Optimizer
+# where we use these optimizer in a functional way.
+# Instead of using the `param.grad` when updating parameters,
+# we explicitly let the user pass gradients to the `step` function
```
Comment (rohan-varma) [inline torch/distributed/optim/functional_adam.py:10]: Just to clarify, so according to the comments below "user" is the DistributedOptimizer API, not RPC application user right? The call to optimizer should remain the same for RPC user?

File: torch/distributed/optim/functional_adam.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        state_steps: List[int] = []
+
+        if len(params) != len(gradients):
+            raise ValueError(
+                "the gradients passed in does not equal to the size of the parameters!"
+                + f"Params length: {len(params)}. "
+                + f"Gradients length: {len(gradients)}"
+            )
+
+        for param, gradient in zip(self.param_group['params'], gradients):
```
Comment (rohan-varma) [inline torch/distributed/optim/functional_adam.py:71]: General question, It looks like the similar code in `torch/optim/adam.py` uses `for p in group['params']`, and then accesses the grad with `p.grad`. Although I'm assuming we can't do this since we need the grads explicitly, since dist autograd doesn't populate p.grad?

**Review by rohan-varma — APPROVED** [review]
LGTM overall, although I mostly compared the changes to `torch/optim/adam.py` and  `torch/distributed/optim/functional_adagrad.py` and checked for parity. I don't have context on the changes in `torch/optim/functional.py` so please get someone to look at that. 

@pritamdamania87 Would be great if you get a chance to take a look at these changes as well.

File: torch/distributed/optim/functional_adam.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                exp_avgs.append(state['exp_avg'])
+                exp_avg_sqs.append(state['exp_avg_sq'])
+
+                if self.amsgrad:
+                    max_exp_avg_sqs.append(state['max_exp_avg_sq'])
+
+                # update the steps for each param group update
+                state['step'] += 1
+                # record the step after step update
+                state_steps.append(state['step'].item())
```
Comment (rohan-varma) [inline torch/distributed/optim/functional_adam.py:99]: I'm guessing all the logic up until the point we call `F.adam` is aiming to emulate `torch/optim/adam.py`, although is there any automated way to guarantee this? Could we dedupe the similar parts into helper functions and call those helper functions here? Alternatively, are we guaranteed that the dist optimizer tests will raise an error if the implementations diverge at all?

File: torch/testing/_internal/distributed/rpc/dist_optimizer_test.py [hunk 4]
```diff
@@ -199,65 +198,12 @@ def test_dist_optim(self):
             self.assertEqual(new_w1, module1.get_w())
             self.assertEqual(new_w2, module2.get_w())
 
+    @dist_init()
+    def test_dist_optim(self):
+        self._test_dist_optim_base(optim.SGD, lr=0.05)
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/rpc/dist_optimizer_test.py:203]: can we just move this to `test_dist_optim_functional`?

File: torch/distributed/optim/functional_adam.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/optim/functional_adam.py:10]: Yes that's right, DistributedOptimizer API actually pass those grads to the step function, let me update the comment to clarify

File: torch/distributed/optim/functional_adam.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/optim/functional_adam.py:71]: yeah in distributed autograd context, we don't populate p.grad, instead we call `dist_autograd.get_gradients(autograd_ctx_id)` to get the list of gradients locally.

File: torch/distributed/optim/functional_adam.py (hunk 3 above)
Comment (wanchaol) [reply torch/distributed/optim/functional_adam.py:99]: yes they are similar indeed, but different across different optimizers as there need to be different states for each optimizer (and this is different from the original adam.py as well bc of TorchScript limitations on some syntax), so it's hard to generalized across them. Though, I think we can guaranteed the implementation will not diverge from the functional part as we shared the computation part, and `test_optim` has a good coverage of it. On the state management side, do you think we should introduce some sort of flag to disable/enable the TorchScript support and compare the results in the test?
